File: java/src/org/openqa/selenium/bidi/script/RemoteValue.java
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.openqa.selenium.bidi.script;

import static java.util.Collections.unmodifiableMap;

import java.io.StringReader;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;
import org.openqa.selenium.json.Json;
import org.openqa.selenium.json.JsonInput;
import org.openqa.selenium.json.TypeToken;

public class RemoteValue {

  private enum Type {
    UNDEFINED("undefined"),
    NULL("null"),
    STRING("string"),
    NUMBER("number"),
    SPECIAL_NUMBER("number"),
    BOOLEAN("boolean"),
    BIGINT("bigint"),
    ARRAY("array"),
    DATE("date"),
    MAP("map"),
    OBJECT("object"),
    REGULAR_EXPRESSION("regexp"),
    SET("set"),
    SYMBOL("symbol"),
    FUNCTION("function"),
    WEAK_MAP("weakmap"),
    WEAK_SET("weakset"),
    ITERATOR("iterator"),
    GENERATOR("generator"),
    ERROR("error"),
    PROXY("proxy"),
    PROMISE("promise"),
    TYPED_ARRAY("typedarray"),
    ARRAY_BUFFER("arraybuffer"),
    NODE_LIST("nodelist"),
    HTML_COLLECTION("htmlcollection"),
    NODE("node"),
    WINDOW("window");

    private final String type;

    Type(String type) {
      this.type = type;
    }

    @Override
    public String toString() {
      return type;
    }

    public static Type findByName(String name) {
      Type result = null;
      for (Type type : values()) {
        if (type.toString().equalsIgnoreCase(name)) {
          result = type;
          break;
        }
      }
      return result;
    }
  }

  private static final Json JSON = new Json();

  private final Type type;

  private final Optional<String> handle;

  private final Optional<Long> internalId;

  private final Optional<Object> value;

  private final Optional<String> sharedId;

  public RemoteValue(
      Type type,
      Optional<String> handle,
      Optional<Long> internalId,
      Optional<Object> value,
      Optional<String> sharedId) {
    this.type = type;
    this.handle = handle;
    this.internalId = internalId;
    this.value = value;
    this.sharedId = sharedId;
  }

  public static RemoteValue fromJson(JsonInput input) {
    Type type = null;

    Optional<String> handle = Optional.empty();

    Optional<Long> internalId = Optional.empty();

    Optional<Object> value = Optional.empty();

    Optional<String> sharedId = Optional.empty();

    input.beginObject();
    while (input.hasNext()) {
      switch (input.nextName()) {
        case "type":
          String typeString = input.read(String.class);
          type = Type.findByName(typeString);
          break;

        case "handle":
          handle = Optional.ofNullable(input.read(String.class));
          break;

        case "internalId":
          internalId = Optional.ofNullable(input.read(Long.class));
          break;

        case "value":
          value = Optional.ofNullable(input.read(Object.class));

          break;

        case "sharedId":
          sharedId = Optional.ofNullable(input.read(String.class));
          break;

        default:
          input.skipValue();
          break;
      }
    }

    input.endObject();

    if (value.isPresent()) {
      value = Optional.ofNullable(deserializeValue(value.get(), type));
    }

    return new RemoteValue(type, handle, internalId, value, sharedId);
  }

  public String getType() {
    return type.toString();
  }

  public Optional<String> getHandle() {
    return handle;
  }

  public Optional<Long> getInternalId() {
    return internalId;
  }

  public Optional<Object> getValue() {
    return value;
  }

  public Optional<String> getSharedId() {
    return sharedId;
  }

  private Map<String, Object> toJson() {
    Map<String, Object> toReturn = new TreeMap<>();

    toReturn.put("type", this.getType());
    handle.ifPresent(handleValue -> toReturn.put("handle", handleValue));
    internalId.ifPresent(id -> toReturn.put("internalId", id));
    value.ifPresent(actualValue -> toReturn.put("value", actualValue));
    sharedId.ifPresent(id -> toReturn.put("sharedId", id));

    return unmodifiableMap(toReturn);
  }

  private static Object deserializeValue(Object value, Type type) {
    Object finalValue;

    switch (type) {
      case ARRAY:
      case NODE_LIST:
      case SET:
        try (StringReader reader = new StringReader(JSON.toJson(value));
            JsonInput input = JSON.newInput(reader)) {
          finalValue = input.read(new TypeToken<List<RemoteValue>>() {}.getType());
        }
        break;

      case MAP:
      case OBJECT:
        List<List<Object>> result = (List<List<Object>>) value;
        Map<Object, RemoteValue> map = new HashMap<>();

        for (List<Object> list : result) {
          Object key = list.get(0);
          if (!(key instanceof String)) {
            try (StringReader reader = new StringReader(JSON.toJson(key));
                JsonInput keyInput = JSON.newInput(reader)) {
              key = keyInput.read(RemoteValue.class);
            }
          }
          try (StringReader reader = new StringReader(JSON.toJson(list.get(1)));
              JsonInput valueInput = JSON.newInput(reader)) {
            RemoteValue value1 = valueInput.read(RemoteValue.class);
            map.put(key, value1);
          }
        }
        finalValue = map;
        break;

      case REGULAR_EXPRESSION:
        try (StringReader reader = new StringReader(JSON.toJson(value));
            JsonInput input = JSON.newInput(reader)) {
          finalValue = input.read(RegExpValue.class);
        }
        break;

      case WINDOW:
        try (StringReader reader = new StringReader(JSON.toJson(value));
            JsonInput input = JSON.newInput(reader)) {
          finalValue = input.read(WindowProxyProperties.class);
        }
        break;

      case NODE:
        try (StringReader reader = new StringReader(JSON.toJson(value));
            JsonInput input = JSON.newInput(reader)) {
          finalValue = input.read(NodeProperties.class);
        }
        break;

      default:
        finalValue = value;
    }

    return finalValue;
  }
}

File: dotnet/src/webdriver/BiDi/Script/RemoteValue.cs
// <copyright file="RemoteValue.cs" company="Selenium Committers">
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
// </copyright>

using System.Text.Json;
using System.Text.Json.Serialization;
using OpenQA.Selenium.BiDi.Json;
using OpenQA.Selenium.BiDi.Json.Converters;

namespace OpenQA.Selenium.BiDi.Script;

// https://github.com/dotnet/runtime/issues/72604
//[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
//[JsonDerivedType(typeof(NumberRemoteValue), "number")]
//[JsonDerivedType(typeof(BooleanRemoteValue), "boolean")]
//[JsonDerivedType(typeof(BigIntRemoteValue), "bigint")]
//[JsonDerivedType(typeof(StringRemoteValue), "string")]
//[JsonDerivedType(typeof(NullRemoteValue), "null")]
//[JsonDerivedType(typeof(UndefinedRemoteValue), "undefined")]
//[JsonDerivedType(typeof(SymbolRemoteValue), "symbol")]
//[JsonDerivedType(typeof(ArrayRemoteValue), "array")]
//[JsonDerivedType(typeof(ObjectRemoteValue), "object")]
//[JsonDerivedType(typeof(FunctionRemoteValue), "function")]
//[JsonDerivedType(typeof(RegExpRemoteValue), "regexp")]
//[JsonDerivedType(typeof(DateRemoteValue), "date")]
//[JsonDerivedType(typeof(MapRemoteValue), "map")]
//[JsonDerivedType(typeof(SetRemoteValue), "set")]
//[JsonDerivedType(typeof(WeakMapRemoteValue), "weakmap")]
//[JsonDerivedType(typeof(WeakSetRemoteValue), "weakset")]
//[JsonDerivedType(typeof(GeneratorRemoteValue), "generator")]
//[JsonDerivedType(typeof(ErrorRemoteValue), "error")]
//[JsonDerivedType(typeof(ProxyRemoteValue), "proxy")]
//[JsonDerivedType(typeof(PromiseRemoteValue), "promise")]
//[JsonDerivedType(typeof(TypedArrayRemoteValue), "typedarray")]
//[JsonDerivedType(typeof(ArrayBufferRemoteValue), "arraybuffer")]
//[JsonDerivedType(typeof(NodeListRemoteValue), "nodelist")]
//[JsonDerivedType(typeof(HtmlCollectionRemoteValue), "htmlcollection")]
//[JsonDerivedType(typeof(NodeRemoteValue), "node")]
//[JsonDerivedType(typeof(WindowProxyRemoteValue), "window")]
[JsonConverter(typeof(RemoteValueConverter))]
public abstract record RemoteValue
{
    public static implicit operator bool(RemoteValue remoteValue) => remoteValue.ConvertTo<bool>();
    public static implicit operator double(RemoteValue remoteValue) => remoteValue.ConvertTo<double>();
    public static implicit operator float(RemoteValue remoteValue) => remoteValue.ConvertTo<float>();
    public static implicit operator int(RemoteValue remoteValue) => remoteValue.ConvertTo<int>();
    public static implicit operator long(RemoteValue remoteValue) => remoteValue.ConvertTo<long>();
    public static implicit operator string?(RemoteValue remoteValue) => remoteValue.ConvertTo<string>();

    public TResult? ConvertTo<TResult>()
        => (this, typeof(TResult)) switch
        {
            (_, Type t) when t.IsAssignableFrom(GetType())
                => (TResult)(object)this,
            (BooleanRemoteValue b, Type t) when t == typeof(bool)
                => (TResult)(object)b.Value,
            (NullRemoteValue, Type t) when !t.IsValueType || Nullable.GetUnderlyingType(t) is not null
                => default,
            (NumberRemoteValue n, Type t) when t == typeof(byte)
                => (TResult)(object)Convert.ToByte(n.Value),
            (NumberRemoteValue n, Type t) when t == typeof(sbyte)
                => (TResult)(object)Convert.ToSByte(n.Value),
            (NumberRemoteValue n, Type t) when t == typeof(short)
                => (TResult)(object)Convert.ToInt16(n.Value),
            (NumberRemoteValue n, Type t) when t == typeof(ushort)
                => (TResult)(object)Convert.ToUInt16(n.Value),
            (NumberRemoteValue n, Type t) when t == typeof(int)
                => (TResult)(object)Convert.ToInt32(n.Value),
            (NumberRemoteValue n, Type t) when t == typeof(uint)
                => (TResult)(object)Convert.ToUInt32(n.Value),
            (NumberRemoteValue n, Type t) when t == typeof(long)
                => (TResult)(object)Convert.ToInt64(n.Value),
            (NumberRemoteValue n, Type t) when t == typeof(ulong)
                => (TResult)(object)Convert.ToUInt64(n.Value),
            (NumberRemoteValue n, Type t) when t == typeof(double)
                => (TResult)(object)n.Value,
            (NumberRemoteValue n, Type t) when t == typeof(float)
                => (TResult)(object)Convert.ToSingle(n.Value),
            (NumberRemoteValue n, Type t) when t == typeof(decimal)
                => (TResult)(object)Convert.ToDecimal(n.Value),
            (StringRemoteValue s, Type t) when t == typeof(string)
                => (TResult)(object)s.Value,
            (ArrayRemoteValue a, Type t) when t.IsArray
                => ConvertRemoteValuesToArray<TResult>(a.Value, t.GetElementType()!),
            (ArrayRemoteValue a, Type t) when t.IsGenericType && t.IsAssignableFrom(typeof(List<>).MakeGenericType(t.GetGenericArguments()[0]))
                => ConvertRemoteValuesToGenericList<TResult>(a.Value, typeof(List<>).MakeGenericType(t.GetGenericArguments()[0])),
            (MapRemoteValue m, Type t) when t.IsGenericType && t.GetGenericArguments().Length == 2 && t.IsAssignableFrom(typeof(Dictionary<,>).MakeGenericType(t.GetGenericArguments()))
                => ConvertRemoteValuesToDictionary<TResult>(m.Value, typeof(Dictionary<,>).MakeGenericType(t.GetGenericArguments())),
            (ObjectRemoteValue o, Type t) when t.IsGenericType && t.GetGenericArguments().Length == 2 && t.IsAssignableFrom(typeof(Dictionary<,>).MakeGenericType(t.GetGenericArguments()))
                => ConvertRemoteValuesToDictionary<TResult>(o.Value, typeof(Dictionary<,>).MakeGenericType(t.GetGenericArguments())),

            (_, Type t) when Nullable.GetUnderlyingType(t) is { } underlying
                => ConvertToNullable<TResult>(underlying),

            _ => throw new InvalidCastException($"Cannot convert {GetType().Name} to {typeof(TResult).FullName}")
        };

    private TResult ConvertToNullable<TResult>(Type underlyingType)
    {
        var convertMethod = typeof(RemoteValue).GetMethod(nameof(ConvertTo))!.MakeGenericMethod(underlyingType);
        var value = convertMethod.Invoke(this, null);
        return (TResult)value!;
    }

    private static TResult ConvertRemoteValuesToArray<TResult>(ImmutableArray<RemoteValue>? remoteValues, Type elementType)
    {
        if (remoteValues is null)
        {
            return (TResult)(object)Array.CreateInstance(elementType, 0);
        }

        var convertMethod = typeof(RemoteValue).GetMethod(nameof(ConvertTo))!.MakeGenericMethod(elementType);
        var items = remoteValues.Value;
        var array = Array.CreateInstance(elementType, items.Length);

        for (int i = 0; i < items.Length; i++)
        {
            var convertedItem = convertMethod.Invoke(items[i], null);
            array.SetValue(convertedItem, i);
        }

        return (TResult)(object)array;
    }

    private static TResult ConvertRemoteValuesToGenericList<TResult>(ImmutableArray<RemoteValue>? remoteValues, Type listType)
    {
        var elementType = listType.GetGenericArguments()[0];
        var list = (System.Collections.IList)Activator.CreateInstance(listType)!;

        if (remoteValues is not null)
        {
            var convertMethod = typeof(RemoteValue).GetMethod(nameof(ConvertTo))!.MakeGenericMethod(elementType);

            foreach (var item in remoteValues.Value)
            {
                var convertedItem = convertMethod.Invoke(item, null);
                list.Add(convertedItem);
            }
        }

        return (TResult)list;
    }

    private static TResult ConvertRemoteValuesToDictionary<TResult>(ImmutableArray<ImmutableArray<RemoteValue>>? remoteValues, Type dictionaryType)
    {
        var typeArgs = dictionaryType.GetGenericArguments();
        var dict = (System.Collections.IDictionary)Activator.CreateInstance(dictionaryType)!;

        if (remoteValues is not null)
        {
            var convertKeyMethod = typeof(RemoteValue).GetMethod(nameof(ConvertTo))!.MakeGenericMethod(typeArgs[0]);
            var convertValueMethod = typeof(RemoteValue).GetMethod(nameof(ConvertTo))!.MakeGenericMethod(typeArgs[1]);

            foreach (var pair in remoteValues.Value)
            {
                if (pair.Length != 2)
                {
                    throw new FormatException($"Expected a pair of RemoteValues for dictionary entry, but got {pair.Length} values.");
                }

                var convertedKey = convertKeyMethod.Invoke(pair[0], null)!;
                var convertedValue = convertValueMethod.Invoke(pair[1], null);
                dict.Add(convertedKey, convertedValue);
            }
        }

        return (TResult)dict;
    }
}

public abstract record PrimitiveProtocolRemoteValue : RemoteValue;

public sealed record NumberRemoteValue([property: JsonConverter(typeof(SpecialNumberConverter))] double Value) : PrimitiveProtocolRemoteValue;

public sealed record BooleanRemoteValue(bool Value) : PrimitiveProtocolRemoteValue;

public sealed record BigIntRemoteValue(string Value) : PrimitiveProtocolRemoteValue;

public sealed record StringRemoteValue(string Value) : PrimitiveProtocolRemoteValue;

public sealed record NullRemoteValue : PrimitiveProtocolRemoteValue;

public sealed record UndefinedRemoteValue : PrimitiveProtocolRemoteValue;

public sealed record SymbolRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record ArrayRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }

    public ImmutableArray<RemoteValue>? Value { get; init; }
}

public sealed record ObjectRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }

    public ImmutableArray<ImmutableArray<RemoteValue>>? Value { get; init; }
}

public sealed record FunctionRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record RegExpRemoteValue(RegExpValue Value) : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record DateRemoteValue(string Value) : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record MapRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }

    public ImmutableArray<ImmutableArray<RemoteValue>>? Value { get; init; }
}

public sealed record SetRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }

    public ImmutableArray<RemoteValue>? Value { get; init; }
}

public sealed record WeakMapRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record WeakSetRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record GeneratorRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record ErrorRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record ProxyRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record PromiseRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record TypedArrayRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record ArrayBufferRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record NodeListRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }

    public ImmutableArray<RemoteValue>? Value { get; init; }
}

public sealed record HtmlCollectionRemoteValue : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }

    public ImmutableArray<RemoteValue>? Value { get; init; }
}

public sealed record NodeRemoteValue(string SharedId, NodeProperties? Value) : RemoteValue, ISharedReference
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

public sealed record WindowProxyRemoteValue(WindowProxyProperties Value) : RemoteValue
{
    public Handle? Handle { get; init; }

    public InternalId? InternalId { get; init; }
}

[JsonConverter(typeof(CamelCaseEnumConverter<Mode>))]
public enum Mode
{
    Open,
    Closed
}


// https://github.com/dotnet/runtime/issues/72604
internal class RemoteValueConverter : JsonConverter<RemoteValue>
{
    public override RemoteValue? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        if (reader.TokenType == JsonTokenType.String)
        {
            return new StringRemoteValue(reader.GetString()!);
        }

        return reader.GetDiscriminator("type") switch
        {
            "number" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<NumberRemoteValue>()),
            "boolean" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<BooleanRemoteValue>()),
            "bigint" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<BigIntRemoteValue>()),
            "string" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<StringRemoteValue>()),
            "null" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<NullRemoteValue>()),
            "undefined" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<UndefinedRemoteValue>()),
            "symbol" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<SymbolRemoteValue>()),
            "array" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<ArrayRemoteValue>()),
            "object" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<ObjectRemoteValue>()),
            "function" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<FunctionRemoteValue>()),
            "regexp" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<RegExpRemoteValue>()),
            "date" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<DateRemoteValue>()),
            "map" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<MapRemoteValue>()),
            "set" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<SetRemoteValue>()),
            "weakmap" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<WeakMapRemoteValue>()),
            "weakset" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<WeakSetRemoteValue>()),
            "generator" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<GeneratorRemoteValue>()),
            "error" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<ErrorRemoteValue>()),
            "proxy" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<ProxyRemoteValue>()),
            "promise" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<PromiseRemoteValue>()),
            "typedarray" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<TypedArrayRemoteValue>()),
            "arraybuffer" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<ArrayBufferRemoteValue>()),
            "nodelist" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<NodeListRemoteValue>()),
            "htmlcollection" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<HtmlCollectionRemoteValue>()),
            "node" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<NodeRemoteValue>()),
            "window" => JsonSerializer.Deserialize(ref reader, options.GetTypeInfo<WindowProxyRemoteValue>()),
            _ => null,
        };
    }

    public override void Write(Utf8JsonWriter writer, RemoteValue value, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }
}
